canonicalwebteam/discourse/parsers.py

```diff
@@ -23,6 +23,214 @@
+    (extra markup around this table doesn't matter)
+    e.g.:
+
+    <h1>Metadata</h1>
+    <details>
+        <summary>Mapping table</summary>
+        <table>
+        <tr><th>Column 1</th><th>Column 2</th></tr>
+        <tr>
+            <td>data 1</td>
+            <td>data 2</td>
+        </tr>
+        <tr>
+            <td>data 3</td>
+            <td>data 4</td>
+        </tr>
+        </table>
+    </details>
+
+    This will typically be generated in Discourse from Markdown similar to
+    the following:
+
+    # Redirects
+
+    [details=Mapping table]
+    | Column 1| Column 2|
+    | -- | -- |
+    | data 1 | data 2 |
+    | data 3 | data 4 |
+
+    The function will return a list of dictionaries of this format:
+    [
+        {"column-1": "data 1", "column-2": "data 2"},
+        {"column-1": "data 3", "column-2": "data 4"},
+    ]
+    """
+    metadata_soup = _get_section(index_soup, "Metadata")
+
+    topics_metadata = []
+    if metadata_soup:
+        titles = [
+            title_soup.text.lower().replace(" ", "_").replace("-", "_")
+            for title_soup in metadata_soup.select("th")
+        ]
+        for row in metadata_soup.select("tr:has(td)"):
+            row_dict = {}
+            for index, value in enumerate(row.select("td")):
+                if value.find("a"):
+                    row_dict["topic_name"] = value.find("a").text
+                row_dict[titles[index]] = "".join(
+                    str(content) for content in value.contents
+                )
+
+            topics_metadata.append(row_dict)
+
+    return topics_metadata
+
+
+def _parse_url_map(index_soup, url_prefix, index_topic_id, url_section_name):
+    """
+    Given the HTML soup of an index topic
+    extract the URL mappings from a "URLs" section.
+
+    This section could be called whatever is
+    passed in `url_section_name` but it must
+    contain a table of
+    "Topic" to "Path" mappings
+    (extra markup around this table doesn't matter)
+    e.g.:
+
+    <h1>URLs</h1>
+    <details>
+        <summary>Mapping table</summary>
+        <table>
+        <tr><th>Topic</th><th>Path</th></tr>
+        <tr>
+            <td><a href="https://forum.example.com/t/page/10">Page</a></td>
+            <td>/cool-page</td>
+        </tr>
+        <tr>
+            <td>
+                <a href="https://forum.example.com/t/place/11">Place</a>
+            </td>
+            <td>/cool-place</td>
+        </tr>
+        </table>
+    </details>
+
+    This will typically be generated in Discourse from Markdown similar to
+    the following:
+
+    # URLs
+
+    [details=Mapping table]
+    | Topic | Path |
+    | -- | -- |
+    | https://forum.example.com/t/place/11| /cool-page |
+    | https://forum.example.com/t/place/11  | /cool-place |
+
+    """
+
+    url_soup = _get_section(index_soup, url_section_name)
+    url_map = {}
+    warnings = []
+
+    if url_soup:
+        for row in url_soup.select("tr:has(td)"):
```

Comment:
I see you've managed to collapse the number of times we're iterating through rows in this way down to only 2 places (`metadata_soup` and `url_soup`). That's nice.

I think you mentioned that the reason you didn't do my suggestion of making a generic `parse_table` function was because in this case the `topic_path` is processed inside the loop with `topic_match`, and so I assume your worry is that if you instead did:

``` python3
for url_map in parse_table(url_soup.select("table")):
     topic_match = TOPIC_URL_MATCH.match(url_map["topic"]["path"])
```

Then you would be looping through the table rows twice, which would be inefficient. Is that right?

I agree what you've done is much tidier, so thank you for that. But if your concern about the above structure is about performance - not looping more than you need - then it would be possible to write the `parse_table` function as a [generator](https://wiki.python.org/moin/Generators) to avoid this issue.

Anyway, I'm not saying you have to do this. What you've already done is already a huge improvement. Just pointing out that it's possible.

Reply:
That's right